circom/aes256_gcm_test.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2023 Jan Lauinger
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package circom
+
+import (
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/test"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/crypto/ecies"
+	"math/rand"
+	"testing"
+	"time"
+)
+
+// AES gcm testing
+type GCM256Wrapper struct {
+	Key          [32]frontend.Variable
+	PlainChunks  []frontend.Variable
+	Iv           [12]frontend.Variable `gnark:",public"`
+	ChunkIndex   frontend.Variable     `gnark:",public"`
+	CipherChunks []frontend.Variable   `gnark:",public"`
+}
+
+// Define declares the circuit's constraints
+func (circuit *GCM256Wrapper) Define(api frontend.API) error {
+
+	aes := NewAES256(api)
+
+	gcm := NewGCM256(api, &aes)
+
+	// verify aes gcm of chunks
+	gcm.Assert256(circuit.Key, circuit.Iv, circuit.ChunkIndex, circuit.PlainChunks, circuit.CipherChunks)
+
+	return nil
+}
+
+type AES interface {
+	Encrypt(key [32]frontend.Variable, pt [16]frontend.Variable) [16]frontend.Variable
+}
+
+func NewGCM256(api frontend.API, aes AES) GCM256 {
+	return GCM256{api: api, aes: aes}
+}
+
+type GCM256 struct {
+	api frontend.API
+	aes AES
+}
+
+// aes gcm encryption
+func (gcm *GCM256) Assert256(key [32]frontend.Variable, iv [12]frontend.Variable, chunkIndex frontend.Variable, plaintext, ciphertext []frontend.Variable) {
```

Comment:
f the plaintext isn't a multiple of 16 bytes (as initially observed), no constraints are imposed.